tensorflow/c/kernels.cc

```diff
@@ -234,10 +331,81 @@ void TF_OpKernelContext_Failure(TF_OpKernelContext* ctx, TF_Status* status) {
     if (s.ok()) {                                                              \
       *val = static_cast<c_type>(v);                                           \
     }                                                                          \
+  }                                                                            \
+  void TF_OpKernelConstruction_GetAttr##func##List(                            \
+      TF_OpKernelConstruction* ctx, const char* attr_name, c_type* vals,       \
+      int max_vals, TF_Status* status) {                                       \
+    TF_SetStatus(status, TF_OK, "");                                           \
+    const auto* attr = GetAttrValue(ctx, attr_name, status);                   \
+    if (!status->status.ok()) return;                                          \
+    if (attr->value_case() != tensorflow::AttrValue::kList) {                  \
+      status->status =                                                         \
+          InvalidArgument("Value for '", attr_name, "' is not a list.");       \
+      return;                                                                  \
+    }                                                                          \
+    const auto len = std::min(max_vals, attr->list().list_field##_size());     \
+    for (int i = 0; i < len; ++i) {                                            \
+      vals[i] = static_cast<c_type>(attr->list().list_field(i));               \
+    }                                                                          \
   }
 
-DEFINE_TF_GETATTR(Type, TF_DataType, tensorflow::DataType)
-DEFINE_TF_GETATTR(Int32, tensorflow::int32, int32_t)
+DEFINE_TF_GETATTR(Type, TF_DataType, tensorflow::DataType, type)
+DEFINE_TF_GETATTR(Int32, int32_t, tensorflow::int32, i)
+DEFINE_TF_GETATTR(Int64, int64_t, tensorflow::int64, i)
+DEFINE_TF_GETATTR(Float, float, float, f)
+DEFINE_TF_GETATTR(Bool, unsigned char, bool, b)
+
+void TF_OpKernelConstruction_GetAttrString(TF_OpKernelConstruction* ctx,
+                                           const char* attr_name, void* value,
+                                           size_t max_length,
+                                           TF_Status* status) {
+  std::string v;
+  auto* cc_ctx = reinterpret_cast<::tensorflow::OpKernelConstruction*>(ctx);
+  ::tensorflow::Status s = cc_ctx->GetAttr(attr_name, &v);
+  ::tensorflow::Set_TF_Status_from_Status(status, s);
+
+  if (!status->status.ok()) return;
+
+  if (max_length <= 0) {
+    return;
+  }
+  std::memcpy(value, v.data(), std::min<size_t>(v.length(), max_length));
```

Comment:
I'm thinking if it is better to return a `TF_StringView`: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/c_api.h;l=131?q=tf_stringview
This probably requires calling https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/node_def_util.h;l=313?q=GetNodeAttrString.

Otherwise, it seems tricky to guess `max_length`. Unless you have a usecase that makes `max_length` approach better.

Reply:
the max_length will be retrieved from  TF_OpKernelConstruction_GetAttrMetadata(simlilar as [ TF_OperationGetAttrStringList](https://github.com/tensorflow/tensorflow/blob/a038cf2695852a649474af55a3ed49f6a72e3965/tensorflow/c/c_api.cc#L1355)),  In our case, we need to malloc an max_length of memory for the string, and pass it to the C API to fill it. TF_StringView is not this use case, from the [comment](https://github.com/tensorflow/tensorflow/blob/a038cf2695852a649474af55a3ed49f6a72e3965/tensorflow/c/c_api.h#L130), the caller does not take the ownership, and not responsible for feeing it, so I thinks it is intend to get from a global object or something like, such as TF_StringView TF_OpKernelConstruction_GetName().
```
// Used to return strings across the C API. The caller does not take ownership
// of the underlying data pointer and is not responsible for freeing it.
typedef struct TF_StringView {
  const char* data;
  size_t len;
} TF_StringView;
```

Reply:
I don't really have a strong opinion on this. So, getting length and passing it in here sounds good the way you have it.

I think `TF_StringView` would work too. Attributes are stored in NodeDef proto: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/node_def.proto;l=64?q=node_def.proto
And this function just gets a reference to it without taking ownership:
https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/node_def_util.h;l=313?q=GetNodeAttrString.
The caller can then copy over to memory they own if needed.